services/motion/builtin/moveonglobe.go

```diff
@@ -0,0 +1,312 @@
+	kb kinematicbase.KinematicBase,
+	componentName string,
+) (motionplan.Plan, error) {
+	inputs, err := kb.CurrentInputs(ctx)
+	if err != nil {
+		return make(motionplan.Plan, 0), err
+	}
+	// TODO: this is really hacky and we should figure out a better place to store this information
+	if len(kb.Kinematics().DoF()) == 2 {
+		inputs = inputs[:2]
+	}
+	planRequest.StartConfiguration = map[string][]referenceframe.Input{componentName: inputs}
+
+	return motionplan.PlanMotion(ctx, planRequest)
+}
+
+func startPollingForReplan(
+	ctx context.Context,
+	period time.Duration,
+	errChan chan error,
+	replanChan chan bool,
+	doneFn func(),
+	fn func(context.Context,
+	) (bool, error),
+) {
+	goutils.ManagedGo(func() {
+		ticker := time.NewTicker(period)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				replan, err := fn(ctx)
+				if err != nil {
+					errChan <- err
+					return
+				}
+				if replan {
+					// TODO: the function that actually monitors obstacles
+					replanChan <- replan
+					return
+				}
+			}
+		}
+	}, doneFn)
+}
+
+func spawnExecute(
+	motionCfg *motion.MotionConfiguration,
+	kb kinematicbase.KinematicBase,
+	ms *builtIn,
+	movementSensor movementsensor.MovementSensor,
+	destination *geo.Point,
+	manager *execManager,
+) {
+	positionPollingPeriod := time.Duration(1000/motionCfg.PositionPollingFreqHz) * time.Millisecond
+	obstaclePollingPeriod := time.Duration(1000/motionCfg.ObstaclePollingFreqHz) * time.Millisecond
+
+	// helper function to manage polling functions
+	// spawn two goroutines that each have the ability to trigger a replan
+	if positionPollingPeriod > 0 {
+		manager.backgroundWorkers.Add(1)
+		startPollingForReplan(
+			manager.cancelCtx,
+			positionPollingPeriod,
+			manager.positionPollingErrChan,
+			manager.replanChan,
+			manager.backgroundWorkers.Done,
+			func(ctx context.Context) (bool, error) {
+				return false, nil
+			})
+	}
+	if obstaclePollingPeriod > 0 {
+		manager.backgroundWorkers.Add(1)
+		startPollingForReplan(
+			manager.cancelCtx,
+			obstaclePollingPeriod,
+			manager.obsticlePollingErrChan,
+			manager.replanChan,
+			manager.backgroundWorkers.Done,
+			func(ctx context.Context,
+			) (bool, error) {
+				return true, nil
+			})
+	}
+
+	// spawn function to execute the plan on the robot
+	manager.backgroundWorkers.Add(1)
+	goutils.ManagedGo(func() {
+		if err := ms.execute(manager.cancelCtx, kb, manager.plan); err != nil {
+			manager.executionErrChan <- err
+			return
+		}
+
+		// the plan has been fully executed so check to see if the GeoPoint we are at is close enough to the goal.
+		success, err := arrivedAtGoal(
+			manager.cancelCtx,
+			movementSensor,
+			destination,
+			motionCfg.PlanDeviationMM,
+		)
+		if err != nil {
+			manager.executionErrChan <- err
+			return
+		}
+
+		if success {
+			manager.successChan <- success
+			return
+		}
+		manager.executionErrChan <- errors.New("failed to arrive at goal")
```

Comment:
IDK if this is how we want to handle this case but if success is not reached & there is no error I think this deadlocks currently